File: Python/153_FindMinimuminRotatedSortedArray.py
class Solution(object):
    def findMin(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """

        h = 0
        t = len(nums)-1

        while nums[h] > nums[t]:
            m = (h + t ) // 2
            if nums[m] > nums[m+1]:
                return nums[m+1]

            if nums[h] > nums[m]:
                t = m
            else:
                h = m

        return nums[0]
        
# if head < tail no rotation
# head > tail: head > middle or middle > tail

File: C++/153_FindMinimumInRotatedSortedArray.cpp
using namespace std;
#include <vector>

class Solution {
public:
    int findMin(vector<int>& nums) {
        int l = 0;
        int r = nums.size() - 1;

        while(l<r) {
            if (nums[l] < nums[r]) return nums[l];
            int m = (l+r) / 2;
            if (nums[r] < nums[m]) l = m+1;
            else r = m;
        }

        return nums[l];
        
    }
};
